src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs

```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
+            return putParameters;
+        }
+
+        private void SetRetention(ServiceDiagnosticSettingsResource properties)
+        {
+            var retentionPolicy = new RetentionPolicy
+            {
+                Enabled = this.RetentionEnabled.Value,
+                Days = this.RetentionInDays.Value
+            };
+
+            if (properties.Logs != null)
             {
-                foreach (var log in properties.Logs)
+                foreach (LogSettings logSettings in properties.Logs)
                 {
-                    log.Enabled = this.Enabled;
+                    logSettings.RetentionPolicy = retentionPolicy;
                 }
+            }
 
-                foreach (var metric in properties.Metrics)
+            if (properties.Metrics != null)
+            {
+                foreach (MetricSettings metricSettings in properties.Metrics)
                 {
-                    metric.Enabled = this.Enabled;
+                    metricSettings.RetentionPolicy = retentionPolicy;
                 }
             }
-            else
+        }
+
+        private void SetSelectedTimegrains(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
             {
-                if (this.Categories != null)
-                {
-                    foreach (string category in this.Categories)
-                    {
-                        LogSettings logSettings = properties.Logs.FirstOrDefault(x => string.Equals(x.Category, category, StringComparison.OrdinalIgnoreCase));
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Timegrains' parameter.");
+            }
 
-                        if (logSettings == null)
-                        {
-                            throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Log category '{0}' is not available", category));
-                        }
+            foreach (string timegrainString in this.Timegrains)
+            {
+                TimeSpan timegrain = XmlConvert.ToTimeSpan(timegrainString);
+                MetricSettings metricSettings = properties.Metrics.FirstOrDefault(x => TimeSpan.Equals(x.TimeGrain, timegrain));
 
-                        logSettings.Enabled = this.Enabled;
-                    }
+                if (metricSettings == null)
+                {
+                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Metric timegrain '{0}' is not available", timegrainString));
                 }
+                metricSettings.Enabled = this.Enabled.Value;
+            }
+        }
 
-                if (this.Timegrains != null)
+        private void SetSelectedCategories(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
+            {
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Categories' parameter.");
```

Comment:
@vcanaa why not create a parameter set where `Enabled` is required if `Categories` is used?